[Coupons] Edit amount

hichamboushaba opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #5536 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
This PR adds support for editing the coupon's amount, and the list of changes is:
1. Add a `hasChanges` property to the ViewModel, and update the Save button depending on its value.
2. Update the amount unit depending on the coupon type.
3. Add some overloads for WCOutlinedTextField to support more advanced scenarios, and a specific component version that allows typed values.
4. Handle the amount edition.

### Testing instructions
1. Make sure the coupons beta toggle is enabled.
2. Open the app, then click on More menu, then coupons.
3. Click on one of the coupons.
4. Click on Menu, then select "edit".
5. Confirm the "Save" button is disabled.
6. Update the content of the amount field.
7. Make sure to test the field properly: handling of cursor position, allowed characters, backspace behavior...
8. Confirm that the "Save" button gets enabled when the amount is different than what the coupon had.

### Images/gif
https://user-images.githubusercontent.com/1657201/167250737-fa40579b-081e-4aab-95eb-02ce540a24de.mov

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hichamboushaba committed:
Add a hasChanges property to the view state

hichamboushaba committed:
Handle the save button state depending on having changes

hichamboushaba committed:
Show the correct amount unit depending on the coupon type

hichamboushaba committed:
Add more implementation of TextFields for different uses in the app

hichamboushaba committed:
Handle coupon amount edit

hichamboushaba committed:
Add some documentation for the text fields

hichamboushaba committed:
Fix detekt issues

hichamboushaba committed:
Add a comment explaining the usage of LaunchedEffect

hichamboushaba committed:
Prevent negative values

hichamboushaba committed:
Fix handling of selection when filtering values

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/component/TextFields.kt:
```diff
@@ -34,9 +45,19 @@ fun WCOutlinedTextField(
     leadingIcon: @Composable (() -> Unit)? = null,
     trailingIcon: @Composable (() -> Unit)? = null,
     isError: Boolean = false,
+    visualTransformation: VisualTransformation = VisualTransformation.None,
+    keyboardOptions: KeyboardOptions = KeyboardOptions.Default,
+    keyboardActions: KeyboardActions = KeyboardActions.Default,
+    singleLine: Boolean = false,
+    maxLines: Int = Int.MAX_VALUE,
+    interactionSource: MutableInteractionSource = remember { MutableInteractionSource() },
```
I noticed that I missed those arguments when I added the component on the previous PR.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/component/TextFields.kt:
```diff
@@ -50,7 +71,156 @@ fun WCOutlinedTextField(
+ */
+@Composable
+fun WCOutlinedTextField(
+    value: TextFieldValue,
```
This version to allow accessing directly the `TextFieldValue` which allows advanced handling (cursor position and selection).

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/component/TextFields.kt:
```diff
@@ -50,7 +71,156 @@ fun WCOutlinedTextField(
+ *        (such as: disallowing empty values, advanced decimal formatting...)
+ */
+@Composable
+fun <T> WCOutlinedTextField(
+    value: T,
+    onValueChange: (T) -> Unit,
+    label: String,
+    parseText: (String) -> T,
+    parseValue: (T) -> String,
+    modifier: Modifier = Modifier,
+    preAdjustText: (String) -> String = { it },
+    helperText: String? = null,
+    enabled: Boolean = true,
+    readOnly: Boolean = false,
+    leadingIcon: @Composable (() -> Unit)? = null,
+    trailingIcon: @Composable (() -> Unit)? = null,
+    isError: Boolean = false,
+    visualTransformation: VisualTransformation = VisualTransformation.None,
+    keyboardOptions: KeyboardOptions = KeyboardOptions.Default,
+    keyboardActions: KeyboardActions = KeyboardActions.Default,
+    singleLine: Boolean = true,
+    maxLines: Int = Int.MAX_VALUE,
+    interactionSource: MutableInteractionSource = remember { MutableInteractionSource() },
+    colors: TextFieldColors = TextFieldDefaults.outlinedTextFieldColors()
+) {
```
Please be nitpicky about this version and its API, if you have any suggestions please share them.
My goal with this is to make it reusable and generic as much as possible, we can then have specific versions for specific use cases if we want, for example we can think about creating a currency text field, you can check the usage in `EditCouponScreen` on how we can consume the API to achieve what we want.

hichamboushaba committed:
Improvements to the typed text field

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/component/TypedTextField.kt:
```diff
@@ -0,0 +1,196 @@
+ * @param valueMapper the [TextFieldValueMapper] to use with this field.
+ */
+@Composable
+fun <T> WCOutlinedTypedTextField(
```
Please be nitpicky about this component and its API (also the interface `TextFieldValueMapper`), if you have any suggestions please share them.

Unlike the View system, the text fields in Compose don't do any filtering (even if you pass `KeyboardType.Number`, the field itself will still accept any entered characters), it's up to the consumer to do it, and this where the  `TextFieldValueMapper` interface comes handy, we can handle both filtering, and parsing the value to a specific type, which improves the communication with the ViewModels.

I tried to make it reusable and generic as much as possible, and with some default implementations for `TextFieldValueMapper` like the ones provided below, we won't have to think much about the fields in different parts of the app, we can even think about an implementation that supports currency formatting.

hichamboushaba committed:
Fix handling of text selection

0nko commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/component/TypedTextField.kt:
```diff
@@ -0,0 +1,200 @@
+ * Handles the mapping between the values of type [T] and their text representation, when used in a
+ * typed [WCOutlinedTypedTextField].
+ */
+interface TextFieldValueMapper<T> {
```
I noticed that going from "0.0" to "0" is considered a change. For `BigDecimal` text fields we should probably consider the numeric equivalence rather than the raw text. Maybe we could add some function like `equals(value: T): Boolean` and let the implementation handle the specifics.

hichamboushaba replied:
Good point, this should be fixed by the commit 60c64748ae438f29e3f720bd45f4a0096e8ab017

hichamboushaba replied:
I spoke too fast, the implementation should be now completed by the additional commit b809402f948f1f4c6376ddda180f9c07aa5562b6, this allows having a custom equality check, while offering a default implementation that supports `Comparable` types.

0nko replied:
Thanks for the update! I tested it but the state of the button still doesn't update as expected. For example, when you start with "0.00", change it to "0.05" and then to "0.0", the button should get disabled because the stored coupon's and draft coupon's amounts are equivalent.

hichamboushaba replied:
This is a bit unrelated to what we fixed here.
What you experienced is because we use the default `equals` function of the `Coupon`, which uses the default `equals` of the `BigDecimal` value.
We can solve this by overriding the `equals` function of `Coupon`, but then we'll lose the auto-generation of the function, which means we need to think about it if we make any changes to the fields, do you think fixing this outweighs the advantages of using the default `equals` function?

hichamboushaba replied:
I added the commit [30eeaa3](https://github.com/woocommerce/woocommerce-android/pull/6462/commits/30eeaa363c3133f7450f4edecd80e466a821f4ce) which fixes this, we'll have to always think about it if we made any changes to the fields, but I don't think there will be a need for this (or at least it'll be rare), so we should be good to go.

Thanks @0nko

0nko replied:
Works great now, thank you!

0nko reviewed (changes requested):
I've played around with the price field and I found a bug with the negative sign character:

1. Delete the price (so that it's 0)
2. Type "-"
3. Start typing any numbers
4. Notice nothing changes
5. Tap at the begging of the text field
6. Notice that the app crashes

This is the stacktrace:
```
java.lang.IllegalArgumentException: start cannot be negative. [start: -26]
        at androidx.compose.ui.text.TextRangeKt.packWithCheck(TextRange.kt:121)
        at androidx.compose.ui.text.TextRangeKt.TextRange(TextRange.kt:37)
        at androidx.compose.ui.text.TextRangeKt.TextRange(TextRange.kt:100)
        at com.woocommerce.android.ui.compose.component.TypedTextFieldKt$WCOutlinedTypedTextField$2.invoke(TypedTextField.kt:76)
        at com.woocommerce.android.ui.compose.component.TypedTextFieldKt$WCOutlinedTypedTextField$2.invoke(TypedTextField.kt:63)
        at androidx.compose.foundation.text.BasicTextFieldKt$BasicTextField$7$1.invoke(BasicTextField.kt:266)
        at androidx.compose.foundation.text.BasicTextFieldKt$BasicTextField$7$1.invoke(BasicTextField.kt:264)
        at androidx.compose.foundation.text.CoreTextFieldKt$CoreTextField$onValueChangeWrapper$1.invoke(CoreTextField.kt:241)
        at androidx.compose.foundation.text.CoreTextFieldKt$CoreTextField$onValueChangeWrapper$1.invoke(CoreTextField.kt:236)
        at androidx.compose.foundation.text.TextFieldDelegate$Companion.setCursorOffset-ULxng0E$foundation_release(TextFieldDelegate.kt:221)
        at androidx.compose.foundation.text.CoreTextFieldKt$CoreTextField$pointerModifier$1.invoke-k-4lQ0M(CoreTextField.kt:299)
        at androidx.compose.foundation.text.CoreTextFieldKt$CoreTextField$pointerModifier$1.invoke(CoreTextField.kt:294)
        at androidx.compose.foundation.text.TextFieldPressGestureFilterKt$tapPressTextFieldModifier$1$2$2.invoke-k-4lQ0M(TextFieldPressGestureFilter.kt:82)
        at androidx.compose.foundation.text.TextFieldPressGestureFilterKt$tapPressTextFieldModifier$1$2$2.invoke(TextFieldPressGestureFilter.kt:55)
        at androidx.compose.foundation.gestures.TapGestureDetectorKt$detectTapAndPress$2$1$1.invokeSuspend(TapGestureDetector.kt:223)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTaskKt.resume(DispatchedTask.kt:178)
        at kotlinx.coroutines.DispatchedTaskKt.dispatch(DispatchedTask.kt:166)
        at kotlinx.coroutines.CancellableContinuationImpl.dispatchResume(CancellableContinuationImpl.kt:397)
        at kotlinx.coroutines.CancellableContinuationImpl.resumeImpl(CancellableContinuationImpl.kt:431)
        at kotlinx.coroutines.CancellableContinuationImpl.resumeImpl$default(CancellableContinuationImpl.kt:420)
        at kotlinx.coroutines.CancellableContinuationImpl.resumeWith(CancellableContinuationImpl.kt:328)
        at androidx.compose.ui.input.pointer.SuspendingPointerInputFilter$PointerEventHandlerCoroutine.offerPointerEvent(SuspendingPointerInputFilter.kt:511)
        at androidx.compose.ui.input.pointer.SuspendingPointerInputFilter.dispatchPointerEvent(SuspendingPointerInputFilter.kt:406)
        at androidx.compose.ui.input.pointer.SuspendingPointerInputFilter.onPointerEvent-H0pRuoY(SuspendingPointerInputFilter.kt:419)
        at androidx.compose.ui.input.pointer.Node.dispatchMainEventPass(HitPathTracker.kt:310)
        at androidx.compose.ui.input.pointer.Node.dispatchMainEventPass(HitPathTracker.kt:297)
        at androidx.compose.ui.input.pointer.Node.dispatchMainEventPass(HitPathTracker.kt:297)
        at androidx.compose.ui.input.pointer.Node.dispatchMainEventPass(HitPathTracker.kt:297)
        at androidx.compose.ui.input.pointer.NodeParent.dispatchMainEventPass(HitPathTracker.kt:179)
        at androidx.compose.ui.input.pointer.HitPathTracker.dispatchChanges(HitPathTracker.kt:98)
        at androidx.compose.ui.input.pointer.PointerInputEventProcessor.process-BIzXfog(PointerInputEventProcessor.kt:80)
        at androidx.compose.ui.platform.AndroidComposeView.sendMotionEvent-8iAsVTc(AndroidComposeView.android.kt:1159)
        at androidx.compose.ui.platform.AndroidComposeView.handleMotionEvent-8iAsVTc(AndroidComposeView.android.kt:1109)
        at androidx.compose.ui.platform.AndroidComposeView.dispatchTouchEvent(AndroidComposeView.android.kt:1059)
        at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:3118)
```

hichamboushaba committed:
Emit change event only when the value itself changes

hichamboushaba committed:
Make sure selection index is set to supported values

hichamboushaba committed:
Fix minus sign handling in Samsung keyboard

hichamboushaba committed:
Add a function to allow using a custom equality check

hichamboushaba commented:
Thanks @0nko for the review.

> I've played around with the price field and I found a bug with the negative sign character:

While I wasn't able to reproduce the exact crash, I was able to make it crash with different steps, I updated the logic, and this should be fixed now (3abc01db8ad85e85c1264429d1d2a17fee8a69be)
I also added a fix (f8e990cc9400955505648ff0b6beecd137cb73f5) for keyboards that use a single button for multiple characters (Samsung keyboard for example, it uses a single button for both `-` and `.` characters).

This is ready for another round.

hichamboushaba committed:
Use safe call instead of double-bang operator

hichamboushaba committed:
Customize the coupon equality check to compare the actual amounts

0nko reviewed (approved):
This looks great now! Nice work! 👍
